### ISSUE TITLE
Connection to external service with ALLOW_ANY Still drop the connection

### ISSUE THREAD

**Issue by AlexandreRoba** [istio/istio#24461]
**Bug description**
I'm deploying a pod that connect to an external service (Postgresql DB) but get a connection error.
```
Unable to obtain connection from database (jdbc:postgresql://XXXXXXXXX:YYYYY/databasename) for user 'username': Connection to XXXXXXXXX:YYYYY refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
```
If disable the side car injection on the pod then the connection works.

Istio was installed with the command:
```
istioctl install --set profile=demo 
```
I found on the documentation here the [Passthrough Envoy configuration](https://istio.io/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services)
When running the command to get the mode it was empty
```
kubectl get configmap istio -n istio-system -o yaml | grep -o "mode: ALLOW_ANY" | uniq
```
So i have reinstalled istio using this command
```
istioctl install --set meshConfig.outboundTrafficPolicy.mode=ALLOW_ANY --set profile=demo 
```
Now when I run the command I'm getting:
```
> $ kubectl get configmap istio -n istio-system -o yaml | grep -o "mode: ALLOW_ANY" | uniq                                                                                                            
mode: ALLOW_ANY
```

But I'm still getting the error :(

Here is the log of the side car proxy:
```
> $ k logs -n dev hophopapi-deployment-6f66cdf5f4-pbx7h istio-proxy                                                                                                                                   
2020-06-05T07:55:41.897697Z	info	FLAG: --concurrency="2"
2020-06-05T07:55:41.897758Z	info	FLAG: --disableInternalTelemetry="false"
2020-06-05T07:55:41.897767Z	info	FLAG: --domain="dev.svc.cluster.local"
2020-06-05T07:55:41.897770Z	info	FLAG: --help="false"
2020-06-05T07:55:41.897774Z	info	FLAG: --id=""
2020-06-05T07:55:41.897777Z	info	FLAG: --ip=""
2020-06-05T07:55:41.897780Z	info	FLAG: --log_as_json="false"
2020-06-05T07:55:41.897783Z	info	FLAG: --log_caller=""
2020-06-05T07:55:41.897786Z	info	FLAG: --log_output_level="default:info"
2020-06-05T07:55:41.897789Z	info	FLAG: --log_rotate=""
2020-06-05T07:55:41.897792Z	info	FLAG: --log_rotate_max_age="30"
2020-06-05T07:55:41.897795Z	info	FLAG: --log_rotate_max_backups="1000"
2020-06-05T07:55:41.897799Z	info	FLAG: --log_rotate_max_size="104857600"
2020-06-05T07:55:41.897802Z	info	FLAG: --log_stacktrace_level="default:none"
2020-06-05T07:55:41.897809Z	info	FLAG: --log_target="[stdout]"
2020-06-05T07:55:41.897813Z	info	FLAG: --meshConfig="./etc/istio/config/mesh"
2020-06-05T07:55:41.897815Z	info	FLAG: --mixerIdentity=""
2020-06-05T07:55:41.897818Z	info	FLAG: --outlierLogPath=""
2020-06-05T07:55:41.897821Z	info	FLAG: --pilotIdentity=""
2020-06-05T07:55:41.897823Z	info	FLAG: --proxyComponentLogLevel="misc:error"
2020-06-05T07:55:41.897826Z	info	FLAG: --proxyLogLevel="warning"
2020-06-05T07:55:41.897829Z	info	FLAG: --serviceCluster="hophopapi.dev"
2020-06-05T07:55:41.897850Z	info	FLAG: --serviceregistry="Kubernetes"
2020-06-05T07:55:41.897853Z	info	FLAG: --stsPort="0"
2020-06-05T07:55:41.897856Z	info	FLAG: --templateFile=""
2020-06-05T07:55:41.897859Z	info	FLAG: --tokenManagerPlugin="GoogleTokenExchange"
2020-06-05T07:55:41.897862Z	info	FLAG: --trust-domain="cluster.local"
2020-06-05T07:55:41.897892Z	info	Version 1.6.0-df8ea82bf434fece015767579ef2e4a16e6a3fb3-dirty-Modified
2020-06-05T07:55:41.898286Z	info	Obtained private IP [100.64.0.106]
2020-06-05T07:55:41.898432Z	info	Apply proxy config from env {"proxyMetadata":{"DNS_AGENT":""}}

2020-06-05T07:55:41.900343Z	info	Effective config: binaryPath: /usr/local/bin/envoy
concurrency: 2
configPath: ./etc/istio/proxy
discoveryAddress: istiod.istio-system.svc:15012
drainDuration: 45s
envoyAccessLogService: {}
envoyMetricsService: {}
parentShutdownDuration: 60s
proxyAdminPort: 15000
proxyMetadata:
  DNS_AGENT: ""
serviceCluster: hophopapi.dev
statNameLength: 189
statusPort: 15020
tracing:
  zipkin:
    address: zipkin.istio-system:9411

2020-06-05T07:55:41.900434Z	info	Proxy role: &model.Proxy{ClusterID:"", Type:"sidecar", IPAddresses:[]string{"100.64.0.106"}, ID:"hophopapi-deployment-6f66cdf5f4-pbx7h.dev", Locality:(*envoy_api_v2_core.Locality)(nil), DNSDomain:"dev.svc.cluster.local", ConfigNamespace:"", Metadata:(*model.NodeMetadata)(nil), SidecarScope:(*model.SidecarScope)(nil), PrevSidecarScope:(*model.SidecarScope)(nil), MergedGateway:(*model.MergedGateway)(nil), ServiceInstances:[]*model.ServiceInstance(nil), IstioVersion:(*model.IstioVersion)(nil), ipv6Support:false, ipv4Support:false, GlobalUnicastIP:"", XdsResourceGenerator:model.XdsResourceGenerator(nil), Active:map[string]*model.WatchedResource(nil)}
2020-06-05T07:55:41.900444Z	info	JWT policy is third-party-jwt
2020-06-05T07:55:41.900505Z	info	PilotSAN []string{"istiod.istio-system.svc"}
2020-06-05T07:55:41.900526Z	info	MixerSAN []string{"spiffe://cluster.local/ns/istio-system/sa/istio-mixer-service-account"}
2020-06-05T07:55:41.900533Z	info	serverOptions.CAEndpoint == istiod.istio-system.svc:15012
2020-06-05T07:55:41.900615Z	info	Using user-configured CA istiod.istio-system.svc:15012
2020-06-05T07:55:41.900619Z	info	istiod uses self-issued certificate
2020-06-05T07:55:41.900748Z	info	the CA cert of istiod is: -----BEGIN CERTIFICATE-----
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-----END CERTIFICATE-----

2020-06-05T07:55:41.900969Z	info	parsed scheme: ""
2020-06-05T07:55:41.900990Z	info	scheme "" not registered, fallback to default scheme
2020-06-05T07:55:41.901021Z	info	ccResolverWrapper: sending update to cc: {[{istiod.istio-system.svc:15012  <nil> 0 <nil>}] <nil> <nil>}
2020-06-05T07:55:41.901034Z	info	ClientConn switching balancer to "pick_first"
2020-06-05T07:55:41.901040Z	info	Channel switches to new LB policy "pick_first"
2020-06-05T07:55:41.901083Z	info	Subchannel Connectivity change to CONNECTING
2020-06-05T07:55:41.901279Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00068c140, {CONNECTING <nil>}
2020-06-05T07:55:41.901338Z	info	Channel Connectivity change to CONNECTING
2020-06-05T07:55:41.901606Z	info	Subchannel picks a new address "istiod.istio-system.svc:15012" to connect
2020-06-05T07:55:41.920881Z	info	Subchannel Connectivity change to READY
2020-06-05T07:55:41.920960Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00068c140, {READY <nil>}
2020-06-05T07:55:41.920970Z	info	Channel Connectivity change to READY
2020-06-05T07:55:41.996112Z	info	sds	SDS gRPC server for workload UDS starts, listening on "./etc/istio/proxy/SDS" 

2020-06-05T07:55:41.996220Z	info	Starting proxy agent
2020-06-05T07:55:41.996459Z	info	Received new config, creating new Envoy epoch 0
2020-06-05T07:55:41.996411Z	info	Opening status port 15020

2020-06-05T07:55:41.996522Z	info	Epoch 0 starting
2020-06-05T07:55:41.996466Z	info	sds	Start SDS grpc server
2020-06-05T07:55:42.008483Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster hophopapi.dev --service-node sidecar~100.64.0.106~hophopapi-deployment-6f66cdf5f4-pbx7h.dev~dev.svc.cluster.local --max-obj-name-len 189 --local-address-ip-version v4 --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error --concurrency 2]
2020-06-05T07:55:42.121813Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamAggregatedResources gRPC config stream closed: 14, no healthy upstream
2020-06-05T07:55:42.121961Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:54] Unable to establish new stream
2020-06-05T07:55:42.146212Z	info	sds	resource:default new connection
2020-06-05T07:55:42.432262Z	info	cache	Root cert has changed, start rotating root cert for SDS clients
2020-06-05T07:55:42.432690Z	info	cache	GenerateSecret default
2020-06-05T07:55:42.433075Z	info	sds	resource:default pushed key/cert pair to proxy
2020-06-05T07:55:43.054196Z	info	sds	resource:ROOTCA new connection
2020-06-05T07:55:43.054776Z	info	cache	Loaded root cert from certificate ROOTCA
2020-06-05T07:55:43.055208Z	info	sds	resource:ROOTCA pushed root cert to proxy
2020-06-05T07:55:43.261825Z	warning	envoy filter	[src/envoy/http/authn/http_filter_factory.cc:83] mTLS PERMISSIVE mode is used, connection can be either plaintext or TLS, and client cert can be omitted. Please consider to upgrade to mTLS STRICT mode for more secure configuration that only allows TLS connection with client cert. See https://istio.io/docs/tasks/security/mtls-migration/
2020-06-05T07:55:43.263476Z	warning	envoy filter	[src/envoy/http/authn/http_filter_factory.cc:83] mTLS PERMISSIVE mode is used, connection can be either plaintext or TLS, and client cert can be omitted. Please consider to upgrade to mTLS STRICT mode for more secure configuration that only allows TLS connection with client cert. See https://istio.io/docs/tasks/security/mtls-migration/
2020-06-05T07:55:44.344937Z	info	Envoy proxy is ready
```
Do not know if this can have an impact but I have also installed on the cluster cert-manager and I use the HTTP01 mode to automatticaly signed the certificate used in the incoming traffic at the gateway.

This is the istio config map
```
> $ k describe configmap -n istio-system istio                                                                                                                                                        
Name:         istio
Namespace:    istio-system
Labels:       install.operator.istio.io/owning-resource=installed-state
              istio.io/rev=default
              operator.istio.io/component=Pilot
              operator.istio.io/managed=Reconcile
              operator.istio.io/version=1.6.0
              release=istio
Annotations:  
Data
====
mesh:
----
accessLogEncoding: TEXT
accessLogFile: /dev/stdout
accessLogFormat: ""
defaultConfig:
  concurrency: 2
  configPath: ./etc/istio/proxy
  connectTimeout: 10s
  controlPlaneAuthPolicy: NONE
  discoveryAddress: istiod.istio-system.svc:15012
  drainDuration: 45s
  parentShutdownDuration: 1m0s
  proxyAdminPort: 15000
  proxyMetadata:
    DNS_AGENT: ""
  serviceCluster: istio-proxy
  tracing:
    zipkin:
      address: zipkin.istio-system:9411
disableMixerHttpReports: true
disablePolicyChecks: false
enablePrometheusMerge: false
ingressClass: istio
ingressControllerMode: STRICT
ingressService: istio-ingressgateway
outboundTrafficPolicy:
  mode: ALLOW_ANY
protocolDetectionTimeout: 100ms
reportBatchMaxEntries: 100
reportBatchMaxTime: 1s
sdsUdsPath: unix:/etc/istio/proxy/SDS
trustDomain: cluster.local
trustDomainAliases: null
meshNetworks:
----
networks: {}
Events:  <none>
```

[ ] Configuration Infrastructure
[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
I expect the connection to work.

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```
> $ istioctl version --remote                                                                                                                                                                         
client version: 1.6.0
control plane version: 1.6.0
data plane version: 1.6.0 (4 proxies)
```

```
> $ kubectl version                                                                                                                                                                                   
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-21T14:51:23Z", GoVersion:"go1.14.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**

```
istioctl install --set meshConfig.outboundTrafficPolicy.mode=ALLOW_ANY --set profile=demo 
```

**Environment where bug was observed (cloud vendor, OS, etc)**

Scaleway.com

**Comment by howardjohn** [comment]
can you grab a config dump from the client pod (https://github.com/istio/istio/wiki/Troubleshooting-Istio) and let us know what port the pg is on?

**Comment by AlexandreRoba** [comment]
@howardjohn thanks for replying to me. Please find here attach the dump of the client pod.

[config_dump.json.zip](https://github.com/istio/istio/files/4737843/config_dump.json.zip)

The port of pg is  44303. Please let me know if you need any other info.

**Comment by howardjohn** [comment]
Weird, I would expect this to be working. thanks for the info. Are there any access logs of the failed requests?

One thing that would be useful is to turn off telemetry and see if there is an impact. An easy way to do this is to remove all the EnvoyFilter's in the istio-system namespace

**Comment by AlexandreRoba** [comment]
Hi @howardjohn, so do I... This is really weird, I already lost several hours on this and do not know what I could do more except to remove the sidecar :( . The logs I have is from the Http Server which is started on the container at the start and then crash right away when migrating the DB with this only log line:
```
Exception in thread "main" org.flywaydb.core.internal.exception.FlywaySqlException: 
Unable to obtain connection from database (jdbc:postgresql://XXX.XXX.XXX.XXX:44303/hophopapidb) for user 'yyyyyyyyyyyy': Connection to XXX.XXXX.XXX.XXX:44303 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL State  : 08001
Error Code : 0
Message    : Connection to XXX.XXX.XXX.XXX:44303 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.

	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:65)
	at org.flywaydb.core.internal.jdbc.JdbcConnectionFactory.<init>(JdbcConnectionFactory.java:80)
	at org.flywaydb.core.Flyway.execute(Flyway.java:456)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:159)
	at services.FlywayService.migrateDatabase(FlywayService.scala:9)
	at WebServer$.delayedEndpoint$WebServer$1(WebServer.scala:13)
	at WebServer$delayedInit$body.apply(WebServer.scala:8)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at scala.Function0.apply$mcV$sp$(Function0.scala:39)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
	at scala.App.$anonfun$main$1(App.scala:73)
	at scala.App.$anonfun$main$1$adapted(App.scala:73)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:920)
	at scala.App.main(App.scala:73)
	at scala.App.main$(App.scala:71)
	at WebServer$.main(WebServer.scala:8)
	at WebServer.main(WebServer.scala)
Caused by: org.postgresql.util.PSQLException: Connection to XXX.XXX.XXX.XXX:44303 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:285)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:211)
	at org.postgresql.Driver.makeConnection(Driver.java:459)
	at org.postgresql.Driver.connect(Driver.java:261)
	at org.flywaydb.core.internal.jdbc.DriverDataSource.getConnectionFromDriver(DriverDataSource.java:474)
	at org.flywaydb.core.internal.jdbc.DriverDataSource.getConnection(DriverDataSource.java:440)
	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:56)
	... 18 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:607)
	at org.postgresql.core.PGStream.<init>(PGStream.java:81)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:93)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:197)
	... 25 more
```

Then I look at the istio side car log, I cannot see any outbound log error from there. I can see couple of warning related to SDS but nothing seems related to the db connection.
```
2020-06-06T08:45:35.186128Z	info	FLAG: --concurrency="2"
2020-06-06T08:45:35.186189Z	info	FLAG: --disableInternalTelemetry="false"
2020-06-06T08:45:35.186200Z	info	FLAG: --domain="dev.svc.cluster.local"
2020-06-06T08:45:35.186204Z	info	FLAG: --help="false"
2020-06-06T08:45:35.186209Z	info	FLAG: --id=""
2020-06-06T08:45:35.186212Z	info	FLAG: --ip=""
2020-06-06T08:45:35.186216Z	info	FLAG: --log_as_json="false"
2020-06-06T08:45:35.186220Z	info	FLAG: --log_caller=""
2020-06-06T08:45:35.186224Z	info	FLAG: --log_output_level="default:info"
2020-06-06T08:45:35.186228Z	info	FLAG: --log_rotate=""
2020-06-06T08:45:35.186232Z	info	FLAG: --log_rotate_max_age="30"
2020-06-06T08:45:35.186237Z	info	FLAG: --log_rotate_max_backups="1000"
2020-06-06T08:45:35.186241Z	info	FLAG: --log_rotate_max_size="104857600"
2020-06-06T08:45:35.186245Z	info	FLAG: --log_stacktrace_level="default:none"
2020-06-06T08:45:35.186357Z	info	FLAG: --log_target="[stdout]"
2020-06-06T08:45:35.186368Z	info	FLAG: --meshConfig="./etc/istio/config/mesh"
2020-06-06T08:45:35.186374Z	info	FLAG: --mixerIdentity=""
2020-06-06T08:45:35.186379Z	info	FLAG: --outlierLogPath=""
2020-06-06T08:45:35.186384Z	info	FLAG: --pilotIdentity=""
2020-06-06T08:45:35.186390Z	info	FLAG: --proxyComponentLogLevel="misc:error"
2020-06-06T08:45:35.186395Z	info	FLAG: --proxyLogLevel="warning"
2020-06-06T08:45:35.186399Z	info	FLAG: --serviceCluster="hophopapi.dev"
2020-06-06T08:45:35.186403Z	info	FLAG: --serviceregistry="Kubernetes"
2020-06-06T08:45:35.186407Z	info	FLAG: --stsPort="0"
2020-06-06T08:45:35.186412Z	info	FLAG: --templateFile=""
2020-06-06T08:45:35.186417Z	info	FLAG: --tokenManagerPlugin="GoogleTokenExchange"
2020-06-06T08:45:35.186421Z	info	FLAG: --trust-domain="cluster.local"
2020-06-06T08:45:35.186560Z	info	Version 1.6.0-df8ea82bf434fece015767579ef2e4a16e6a3fb3-dirty-Modified
2020-06-06T08:45:35.186850Z	info	Obtained private IP [100.64.0.192]
2020-06-06T08:45:35.187087Z	info	Apply proxy config from env {"proxyMetadata":{"DNS_AGENT":""}}

2020-06-06T08:45:35.189336Z	info	Effective config: binaryPath: /usr/local/bin/envoy
concurrency: 2
configPath: ./etc/istio/proxy
discoveryAddress: istiod.istio-system.svc:15012
drainDuration: 45s
envoyAccessLogService: {}
envoyMetricsService: {}
parentShutdownDuration: 60s
proxyAdminPort: 15000
proxyMetadata:
  DNS_AGENT: ""
serviceCluster: hophopapi.dev
statNameLength: 189
statusPort: 15020
tracing:
  zipkin:
    address: zipkin.istio-system:9411

2020-06-06T08:45:35.189424Z	info	Proxy role: &model.Proxy{ClusterID:"", Type:"sidecar", IPAddresses:[]string{"100.64.0.192"}, ID:"hophopapi-deployment-669f69d468-hhhdj.dev", Locality:(*envoy_api_v2_core.Locality)(nil), DNSDomain:"dev.svc.cluster.local", ConfigNamespace:"", Metadata:(*model.NodeMetadata)(nil), SidecarScope:(*model.SidecarScope)(nil), PrevSidecarScope:(*model.SidecarScope)(nil), MergedGateway:(*model.MergedGateway)(nil), ServiceInstances:[]*model.ServiceInstance(nil), IstioVersion:(*model.IstioVersion)(nil), ipv6Support:false, ipv4Support:false, GlobalUnicastIP:"", XdsResourceGenerator:model.XdsResourceGenerator(nil), Active:map[string]*model.WatchedResource(nil)}
2020-06-06T08:45:35.189456Z	info	JWT policy is third-party-jwt
2020-06-06T08:45:35.189540Z	info	PilotSAN []string{"istiod.istio-system.svc"}
2020-06-06T08:45:35.189559Z	info	MixerSAN []string{"spiffe://cluster.local/ns/istio-system/sa/istio-mixer-service-account"}
2020-06-06T08:45:35.189569Z	info	serverOptions.CAEndpoint == istiod.istio-system.svc:15012
2020-06-06T08:45:35.189575Z	info	Using user-configured CA istiod.istio-system.svc:15012
2020-06-06T08:45:35.189579Z	info	istiod uses self-issued certificate
2020-06-06T08:45:35.189656Z	info	the CA cert of istiod is: -----BEGIN CERTIFICATE-----
XXXXXXX
-----END CERTIFICATE-----

2020-06-06T08:45:35.189912Z	info	parsed scheme: ""
2020-06-06T08:45:35.189944Z	info	scheme "" not registered, fallback to default scheme
2020-06-06T08:45:35.189974Z	info	ccResolverWrapper: sending update to cc: {[{istiod.istio-system.svc:15012  <nil> 0 <nil>}] <nil> <nil>}
2020-06-06T08:45:35.189999Z	info	ClientConn switching balancer to "pick_first"
2020-06-06T08:45:35.190009Z	info	Channel switches to new LB policy "pick_first"
2020-06-06T08:45:35.190060Z	info	Subchannel Connectivity change to CONNECTING
2020-06-06T08:45:35.190393Z	info	Subchannel picks a new address "istiod.istio-system.svc:15012" to connect
2020-06-06T08:45:35.192672Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc000778020, {CONNECTING <nil>}
2020-06-06T08:45:35.192737Z	info	Channel Connectivity change to CONNECTING
2020-06-06T08:45:35.212682Z	info	Subchannel Connectivity change to READY
2020-06-06T08:45:35.213123Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc000778020, {READY <nil>}
2020-06-06T08:45:35.213227Z	info	Channel Connectivity change to READY
2020-06-06T08:45:35.268096Z	info	sds	SDS gRPC server for workload UDS starts, listening on "./etc/istio/proxy/SDS" 

2020-06-06T08:45:35.268509Z	info	Starting proxy agent
2020-06-06T08:45:35.268969Z	info	Opening status port 15020

2020-06-06T08:45:35.269313Z	info	Received new config, creating new Envoy epoch 0
2020-06-06T08:45:35.269560Z	info	Epoch 0 starting
2020-06-06T08:45:35.268472Z	info	sds	Start SDS grpc server
2020-06-06T08:45:35.284718Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster hophopapi.dev --service-node sidecar~100.64.0.192~hophopapi-deployment-669f69d468-hhhdj.dev~dev.svc.cluster.local --max-obj-name-len 189 --local-address-ip-version v4 --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error --concurrency 2]
2020-06-06T08:45:35.392702Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamAggregatedResources gRPC config stream closed: 14, no healthy upstream
2020-06-06T08:45:35.392855Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:54] Unable to establish new stream
2020-06-06T08:45:35.410842Z	info	sds	resource:default new connection
2020-06-06T08:45:35.627243Z	info	cache	Root cert has changed, start rotating root cert for SDS clients
2020-06-06T08:45:35.627585Z	info	cache	GenerateSecret default
2020-06-06T08:45:35.628097Z	info	sds	resource:default pushed key/cert pair to proxy
2020-06-06T08:45:36.165434Z	info	sds	resource:ROOTCA new connection
2020-06-06T08:45:36.165633Z	info	cache	Loaded root cert from certificate ROOTCA
2020-06-06T08:45:36.165756Z	info	sds	resource:ROOTCA pushed root cert to proxy
2020-06-06T08:45:36.359546Z	warning	envoy filter	[src/envoy/http/authn/http_filter_factory.cc:83] mTLS PERMISSIVE mode is used, connection can be either plaintext or TLS, and client cert can be omitted. Please consider to upgrade to mTLS STRICT mode for more secure configuration that only allows TLS connection with client cert. See https://istio.io/docs/tasks/security/mtls-migration/
2020-06-06T08:45:36.362624Z	warning	envoy filter	[src/envoy/http/authn/http_filter_factory.cc:83] mTLS PERMISSIVE mode is used, connection can be either plaintext or TLS, and client cert can be omitted. Please consider to upgrade to mTLS STRICT mode for more secure configuration that only allows TLS connection with client cert. See https://istio.io/docs/tasks/security/mtls-migration/
2020-06-06T08:45:37.985035Z	info	Envoy proxy is ready
[2020-06-06T08:48:53.013Z] "GET /professionals/ HTTP/1.1" 503 UF "-" "-" 0 91 0 - "XXX.XXX.XXX.XXX" "PostmanRuntime/7.25.0" "39d342b8-87a1-98f5-8590-07598568b368" "XXXXXXXX.com" "127.0.0.1:9000" inbound|9000|http|hophopapi-service.dev.svc.cluster.local - 100.64.0.192:9000 XXX.XXX.XXX.XXX:0 outbound_.9000_._.hophopapi-service.dev.svc.cluster.local default
[2020-06-06T08:48:53.062Z] "GET /professionals/ HTTP/1.1" 503 UF "-" "-" 0 91 0 - "XXX.XXX.XXX.XXX" "PostmanRuntime/7.25.0" "39d342b8-87a1-98f5-8590-07598568b368" "XXXXXXXX.com" "127.0.0.1:9000" inbound|9000|http|hophopapi-service.dev.svc.cluster.local - 100.64.0.192:9000 XXX.XXX.XXX.XXX:0 outbound_.9000_._.hophopapi-service.dev.svc.cluster.local default
[2020-06-06T08:48:53.075Z] "GET /professionals/ HTTP/1.1" 503 UF "-" "-" 0 91 0 - "XXX.XXX.XXX.XXX" "PostmanRuntime/7.25.0" "39d342b8-87a1-98f5-8590-07598568b368" "XXXXXXXX.com" "127.0.0.1:9000" inbound|9000|http|hophopapi-service.dev.svc.cluster.local - 100.64.0.192:9000 XXX.XXX.XXX.XXX:0 outbound_.9000_._.hophopapi-service.dev.svc.cluster.local default
```
I'll try to remove all telemtry and post right back.

**Comment by AlexandreRoba** [comment]
Hi @howardjohn, I have upgraded to 1.6.1 but still I cannot connect. 

I could not find any filters to remove if you have some links that explains this, I'll be glad to have them. 

I tried to create a service entry as decribed [here](https://istio.io/docs/tasks/traffic-management/egress/egress-control/#access-an-external-http-service). Not sure if this is applicable to TCP service, but it did not resolve the issue:
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: postgres-ext
  namespace: dev
spec:
  hosts:
  - xxxxxxxxx.com
  ports:
  - number: 44303
    name: tcp
    protocol: TCP
  resolution: DNS
  location: MESH_EXTERNAL
``` 
Any other idea?

**Comment by AlexandreRoba** [comment]
I have remove all sidecar injection by commenting  #istio-injection: enabled on my namespace, reinstall charts and now connection are working as expected...
I would have never though that the most standard installation with a istioctl install set profile=demo and a single pod that connect to an external DB would give so much difficulties :(

**Comment by howardjohn** [comment]
Is this failing just at startup in the app? It may be because Envoy is not yet running by the time the app makes the request?

For the telemetry run `kubectl delete envoyfilters --all -n istio-system`

**Comment by AlexandreRoba** [comment]
Hi @howardjohn, thanks for the idea. I added a thread sleep of 5 seconds before connecting to the DB. unfortunattely still no success:
```
Waiting before calling the DB
Migrating the db
Exception in thread "main" org.flywaydb.core.internal.exception.FlywaySqlException: 
Unable to obtain connection from database (jdbc:postgresql://xxxxxxxxx.com:44303/hophopapidb) for user 'yyyyyyy': Connection to xxxxxxxx.com:44303 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
I will try to delete the filters.

**Comment by AlexandreRoba** [comment]
I have deleted all the filters and redeploy the app
```
> $ kubectl delete envoyfilters --all -n istio-system                                                                                                                              
envoyfilter.networking.istio.io "metadata-exchange-1.4" deleted
envoyfilter.networking.istio.io "metadata-exchange-1.5" deleted
envoyfilter.networking.istio.io "metadata-exchange-1.6" deleted
envoyfilter.networking.istio.io "stats-filter-1.4" deleted
envoyfilter.networking.istio.io "stats-filter-1.5" deleted
envoyfilter.networking.istio.io "stats-filter-1.6" deleted
envoyfilter.networking.istio.io "tcp-metadata-exchange-1.5" deleted
envoyfilter.networking.istio.io "tcp-metadata-exchange-1.6" deleted
envoyfilter.networking.istio.io "tcp-stats-filter-1.5" deleted
envoyfilter.networking.istio.io "tcp-stats-filter-1.6" deleted
```

And now finally it worked :)

``` 
Waiting before calling the DB
Migrating the db
Database migrated
Server listening on 0.0.0.0:9000
```

**Comment by AlexandreRoba** [comment]
Now I need to look in the doc to understand what those filters do :)

**Comment by howardjohn** [comment]
What they should not be doing is breaking this traffic - I think that is a bug

@gargnupur @mandarjog

**Comment by nataraj24** [comment]
Hi Team,

i am facing similar issue even after deleting envoy filters.   Getting below error while connecting external db.

SQLException encountered inside getDBConnection method java.sql.SQLException: Cannot create PoolableConnectionFactory (The connection attempt failed.)

any resolution would be helpful

**Comment by AlexandreRoba** [comment]
@nataraj24 Yesterday after reinstalling the charts (which includes the pod that connect to the database) I had the connection refused again. I needed to add a delay right after starting the service and the connection to the DB. It now seems to work every time the pod is started.

**Comment by gargnupur** [comment]
@AlexandreRoba : Do you still have all the filters removed to make it work?
Would it be possible to turn on debug logging on the pod? Also, can you just delete these 2 tcp filters: 
1.  "tcp-metadata-exchange-1.5"
2. "tcp-metadata-exchange-1.6" 
and see if the issue still repro's

@mandarjog : For connection to external services, ALPN should not match and thus metadata exchange filter should be a no-op, so this is weird..

**Comment by howardjohn** [comment]
@AlexandreRoba can you clarify if you are seeing this issue with the EnvoyFilters or without? Based on your last comment its not clear if the issue is the telemetry or just https://github.com/istio/istio/issues/11130

**Comment by nataraj24** [comment]
The same image works fine with istio 1.5.2 without introducing any delay or any change.  The external DB connection issue occurs with and without EnvoyFilters in both istio 1.6.1 and 1.6.0.  Below is the full stack trace from application pod and outbound logs of istio-proxy side car.  Request your help and resolution on this 

2020-06-09T03:45:28.975361Z     info    Epoch 0 starting
2020-06-09T03:45:28.997182Z     info    Subchannel Connectivity change to READY
2020-06-09T03:45:29.002770Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000b1eec0, {READY <nil>}
2020-06-09T03:45:29.003084Z     info    Channel Connectivity change to READY

[2020-06-09T03:46:49.878Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52588 - -
[2020-06-09T03:46:58.378Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52636 - -
[2020-06-09T03:46:58.681Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52640 - -
[2020-06-09T03:46:58.779Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52642 - -
[2020-06-09T03:46:59.172Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52648 - -
[2020-06-09T03:46:59.175Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52650 - -
[2020-06-09T03:46:59.177Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52652 - -
[2020-06-09T03:46:59.179Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52654 - -
[2020-06-09T03:46:59.181Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52656 - -
[2020-06-09T03:46:59.271Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52658 - -
[2020-06-09T03:46:59.273Z] "- - -" 0 UH "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 10.99.132.118:6501 10.244.1.19:52660 - -

java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:210) ~[?:1.8.0_232]
        at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[?:1.8.0_232]
        at com.edb.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:144) ~[edb-jdbc17.jar:?]
        at com.edb.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:113) ~[edb-jdbc17.jar:?]
        at com.edb.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:72) ~[edb-jdbc17.jar:?]
        at com.edb.core.PGStream.ReceiveChar(PGStream.java:307) ~[edb-jdbc17.jar:?]
        at com.edb.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:437) ~[edb-jdbc17.jar:?]
        at com.edb.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:197) ~[edb-jdbc17.jar:?]
        at com.edb.core.ConnectionFactory.openConnection(ConnectionFactory.java:67) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc2.AbstractJdbc2Connection.<init>(AbstractJdbc2Connection.java:132) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc3.AbstractJdbc3Connection.<init>(AbstractJdbc3Connection.java:30) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc3g.AbstractJdbc3gConnection.<init>(AbstractJdbc3gConnection.java:22) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc4.AbstractJdbc4Connection.<init>(AbstractJdbc4Connection.java:47) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc4.Jdbc4Connection.<init>(Jdbc4Connection.java:25) ~[edb-jdbc17.jar:?]
        at com.edb.Driver.makeConnection(Driver.java:413) ~[edb-jdbc17.jar:?]
        at com.edb.Driver.connect(Driver.java:285) ~[edb-jdbc17.jar:?]
        at org.apache.tomcat.dbcp.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:55) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:355) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:115) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:665) ~[tomcat-dbcp.jar:9.0.34]
        ... 5 more
Wrapped by: com.edb.util.PSQLException: The connection attempt failed.
        at com.edb.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:288) ~[edb-jdbc17.jar:?]
        at com.edb.core.ConnectionFactory.openConnection(ConnectionFactory.java:67) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc2.AbstractJdbc2Connection.<init>(AbstractJdbc2Connection.java:132) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc3.AbstractJdbc3Connection.<init>(AbstractJdbc3Connection.java:30) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc3g.AbstractJdbc3gConnection.<init>(AbstractJdbc3gConnection.java:22) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc4.AbstractJdbc4Connection.<init>(AbstractJdbc4Connection.java:47) ~[edb-jdbc17.jar:?]
        at com.edb.jdbc4.Jdbc4Connection.<init>(Jdbc4Connection.java:25) ~[edb-jdbc17.jar:?]
        at com.edb.Driver.makeConnection(Driver.java:413) ~[edb-jdbc17.jar:?]
        at com.edb.Driver.connect(Driver.java:285) ~[edb-jdbc17.jar:?]
        at org.apache.tomcat.dbcp.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:55) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:355) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:115) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:665) ~[tomcat-dbcp.jar:9.0.34]
        ... 5 more
Wrapped by: java.sql.SQLException: Cannot create PoolableConnectionFactory (The connection attempt failed.)
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:669) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:544) ~[tomcat-dbcp.jar:9.0.34]
        at org.apache.tomcat.dbcp.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:753) ~[tomcat-dbcp.jar:9.0.34]

**Comment by robincher** [comment]
I encountered something similar, and what i did was to restart all the pods and somehow the outbound policy are finally being applied.

I suspect that sometimes, the pods are not properly inject with Istio sidecar upon start-up.

**Comment by AlexandreRoba** [comment]
Hi guys, I'm still facing connection drops at startup event with the delay. It is random. Sometime the deployment works some time it does not. I have to uninstall and reinstall untill it works... :(

**Comment by AlexandreRoba** [comment]
The clarification is valid. Restarting the pod solve the issue.
I have disabled the sidecar injection for the moment. I will try to get the debug log next wednesday.

**Comment by sameerbhadouria** [comment]
I am experiencing something similar with Redis. I have a script in my app container to check for Istio proxy being available before starting as suggested https://github.com/istio/istio/issues/11130
Had created an separate issue here: https://github.com/istio/istio/issues/25379 but this seems related.

**Comment by nataraj24** [comment]
Hi @howardjohn 

Is this issue fixed in istio 1.7?  The same issue persists in 1.6.7 also, connection to external service doesn't work.  

Thanks,
Nataraj

**Comment by AlexandreRoba** [comment]
> I am experiencing something similar with Redis. I have a script in my app container to check for Istio proxy being available before starting as suggested #11130
> Had created a separate issue here: #25379 but this seems related.
> 
> @AlexandreRoba have you left sidecar injection disabled or were you able to resolve this issue?

Hi sorry for the late reply. I did left side car injection turnoff. I finaly got some time today to start investigating again this "issue". Unfortunately I still have it on the latest version and on two different cluster :( I will keep the side car injection disabled untill this is resolved. It has been like this for several months :(

**Comment by nataraj24** [comment]
this external service issue persists from 1.5.2 on wards till latest version 1.7.0

**Comment by AlexandreRoba** [comment]
@nataraj24 I confirm this is still the case on 1.7... It is even wroste. Now I cannot connect at all to my DB as Service. Previously I would simply kill the pod and then after restarting the pod was connecting... Now I get the error all the time :(

**Comment by satyendrakarn** [comment]
Hi I am facing the same problem with istio 1.7.4. @AlexandreRoba Were you able to work through it? Any pointers would be much appreciated.

**Comment by AlexandreRoba** [comment]
@satyendrakarn Unfortunattely no. I have to restart the pods accessing external ressource when using the proxy sidecar in order for the connection to work. I gave up... I remove the proxy on pods accessing external resources... :(

**Comment by satyendrakarn** [comment]
@AlexandreRoba ok. Can you reopen this issue?

**Comment by schue** [comment]
We seem to be seeing a similar problem with PostgreSQL and Postgraphile. The log shows Postgraphile trying to reconnect but never recovers. Recycling the Postgraphile pod restores functionality.

**Comment by ltmleo** [comment]
Hi, same problem here.

**istio: 1.6.3**

Random pods cannot connect in some external endpoints.

The annotaion: 'traffic.sidecar.istio.io/includeOutboundIPRanges' showing that is a sidecar problem. 

I've noticied that the sidecar logs, don't show allo_any, when this error ocurrs. 
```
[timestamp] "GET /endpoint HTTP /1.1" 404 NR "-" "-" 0 0 0 - "-" "curl/7.64.0" "somehash" "DNS" "-" - - IP:80 IP:35368 - -  >> **NOK**
 [timestamp] "GET /endpoint HTTP /1.1" 401 - "-" "-" 0 128 56 55 "-" "curl/7.64.0" "somehash" "DNS" "IP:80" PassthroughCluster IP:52092 IP:80 IP:52090 - allow_any >> **OK**
```

They have the same configuration and are in the same namespace, I made the request using curl inside the pod.
